mangaki/mangaki/static/js/vote.js

```diff
@@ -1,60 +1,330 @@
-                $(elt).removeClass('not-chosen');
-            }
-        }
+/* Wraps a Promise to ensure that it will resolve with the correct animations.
+ * This returns a promise with the same result as the given promise, but it
+ * guarantees that it will only resolve when the card is dehydrated (hidden)
+ * and takes care of properly displaying transitions (including showing the
+ * loading screen if needed) while waiting for the promise to resolve.
+ *
+ * Technically, `parallelLoad` does something similar to the following:
+ *
+ *  - If the card is currently showing the loading screen, we simply wait for
+ *  the promise to resolve, then dehydrate the card and return the promise's
+ *  result.
+ *  - If the card is currently hydrated with a non-empty work, we first
+ *  dehydrate the card, then check whether the promise finished. If that is the
+ *  case, we return the promise's result; otherwise we re-hydrate the card with
+ *  an empty work (i.e., display the loading screen), wait for the promise to
+ *  resolve, then return the promise's result.
+ */
+Card.prototype.parallelLoad = function (promise) {
+  var card = this;
+  // Ensure we are actually handling a Promise.
+  promise = Promise.resolve(promise);
+
+  if (this.$el.hasClass('work-card_loading')) {
+    return promise.then(function (work) {
+      return card.dehydrate().then(function () {
+        return work;
+      });
     });
+  }
+
+  // FIXME: We really should use a Symbol here.
+  var dehydrate = this.dehydrate().then(function () { return undefined; });
+
+  return Promise.race([dehydrate, promise]).then(function (work) {
+    if (typeof work === 'undefined') {
+      return Promise.all([promise, card.hydrate()]).then(function (results) {
+	var work = results[0];
+
+	return card.dehydrate().then(function () {
+	  return work;
+	});
+      });
+    } else {
+      return dehydrate.then(function () {
+	return work;
+      });
+    }
+  });
+}
+
+/* A Slot is a cache around an endpoint returning works in JSON format as
+ * expected by `Card.hydrate`. It provides a simple, Promise-based API for
+ * getting the next cached work as well as requesting additional work objects
+ * from the given URL on the server.
+ */
+function Slot(url) {
+  this.url = url;
+  this.works = [];
+
+  // Prefetch works
+  this.fetch();
+
+  return this;
+}
+
+/* Returns additional work objects from the server. The Slot uses internal
+ * variables to ensure that there never are two requests from the same Slot to
+ * the server, and that any caller that requested a `fetch()` during a single
+ * request was in-flight will get their promises resolved.
+ *
+ * Returns a promise for the slots' new list of works after getting fetched
+ * from the server.
+ */
+Slot.prototype.fetch = function () {
+  if (typeof this.fetcher !== 'undefined') {
+    return this.fetcher;
+  }
+
+  var slot = this, works = this.works;
+  return this.fetcher = Promise.resolve($.getJSON(this.url)).catch(function (exn) {
```

Comment:
Nope, `return x = y;` is a shortcut for `x = y; return x;` (in this case it's actually slightly faster because there is no attribute access). It's also why `x = y = z` works -- nothing specific to JS, it works the same way in most languages derived from C such as C, C++, Java, PHP, etc.